File: archiver/src/main/scala/farjs/archiver/ArchiverPluginUi.scala
package farjs.archiver

import farjs.archiver.ArchiverPluginUi._
import farjs.filelist.FileListActions._
import farjs.filelist._
import farjs.filelist.api.{FileListDir, FileListItem}
import scommons.react._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.scalajs.js

class ArchiverPluginUi(data: FileListData, archName: String, archType: String, addToArchApi: js.Function4[String, String, js.Set[String], js.Function0[Unit], js.Promise[Unit]], items: js.Array[FileListItem])
    extends FunctionComponent[FileListPluginUiProps] {

  protected def render(compProps: Props): ReactElement = {
    val props = compProps.plain
    
    <(addToArchController)(^.plain := AddToArchControllerProps(
      dispatch = data.dispatch,
      actions = data.actions,
      state = data.state,
      archName = archName,
      archType = archType,
      archAction = AddToArchAction.Add,
      addToArchApi = addToArchApi,
      items = items,
      onComplete = { archFile =>
        props.onClose()

        val action = data.actions.updateDir(data.dispatch, data.state.currDir.path)
        data.dispatch(action)
        action.task.result.toFuture.foreach { updatedDir =>
          data.dispatch(FileListItemCreatedAction(archFile, updatedDir.asInstanceOf[FileListDir]))
        }
      },
      onCancel = props.onClose
    ))()
  }
}

object ArchiverPluginUi {

  private[archiver] var addToArchController: ReactClass = AddToArchController
}

File: archiver/ArchiverPluginUi.mjs
/**
 * @import { FileListItem } from "@farjs/filelist/api/FileListItem.mjs"
 * @import { FileListData } from "@farjs/filelist/FileListData.mjs"
 * @import { FileListItemCreatedAction } from "@farjs/filelist/FileListActions.mjs"
 * @import { FileListPluginUiProps } from "@farjs/filelist/FileListPlugin.mjs"
 */
import React from "react";
import AddToArchController from "./AddToArchController.mjs";

const h = React.createElement;

/**
 * @typedef {{
 *  readonly data: FileListData;
 *  readonly archName: string;
 *  readonly archType: string;
 *  addToArchApi(zipFile: string, parent: string, items: Set<string>, onNextItem: () => void): Promise<void>;
 *  readonly items: readonly FileListItem[];
 * }} ArchiverPluginUiParams
 */

/**
 * @param {ArchiverPluginUiParams} params
 */
function ArchiverPluginUi({ data, archName, archType, addToArchApi, items }) {
  /**
   * @param {FileListPluginUiProps} props
   */
  const ArchiverPluginUiComp = (props) => {
    const { addToArchController } = ArchiverPluginUi;

    return h(addToArchController, {
      dispatch: data.dispatch,
      actions: data.actions,
      state: data.state,
      archName,
      archType,
      archAction: "Add",
      addToArchApi: addToArchApi,
      items,
      onComplete: (archFile) => {
        props.onClose();

        const action = data.actions.updateDir(
          data.dispatch,
          data.state.currDir.path
        );
        data.dispatch(action);
        action.task.result.then((updatedDir) => {
          /** @type {FileListItemCreatedAction} */
          const action = {
            action: "FileListItemCreatedAction",
            name: archFile,
            currDir: updatedDir,
          };
          data.dispatch(action);
        });
      },
      onCancel: props.onClose,
    });
  };

  ArchiverPluginUiComp.displayName = "ArchiverPluginUi";

  return ArchiverPluginUiComp;
}

ArchiverPluginUi.addToArchController = AddToArchController;

export default ArchiverPluginUi;
